ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java

```diff
@@ -260,17 +273,28 @@ public void verifyPrivacyGroupContainsEnclavePublicKey(
   public void verifyPrivacyGroupContainsEnclavePublicKey(
       final String privacyGroupId, final String enclavePublicKey, final Optional<Long> blockNumber)
       throws MultiTenancyValidationException {
-    // TODO: There's potentially a bug here where an onchain privacyGroup
-    // that isn't found will default to the offchain privacy group.
-    final PrivacyGroup privacyGroup =
-        onchainPrivacyGroupContract
-            .flatMap(
-                contract -> contract.getPrivacyGroupByIdAndBlockNumber(privacyGroupId, blockNumber))
-            .orElse(enclave.retrievePrivacyGroup(privacyGroupId));
-
-    if (!privacyGroup.getMembers().contains(enclavePublicKey)) {
-      throw new MultiTenancyValidationException(
-          "Privacy group must contain the enclave public key");
+    Optional<PrivacyGroup> maybePrivacyGroup;
+    PrivacyGroup offchainPrivacyGroup;
+    if (onchainPrivacyGroupContract.isPresent()) {
+      maybePrivacyGroup =
+          onchainPrivacyGroupContract
+              .get()
+              .getPrivacyGroupByIdAndBlockNumber(privacyGroupId, blockNumber);
+      if (maybePrivacyGroup.isEmpty()) {
+        // group doesn't exist yet - this is normal for onchain privacy groups
+      } else {
+        // group DOES exist - does it contain the user?
+        if (!maybePrivacyGroup.get().getMembers().contains(enclavePublicKey)) {
+          throw new MultiTenancyValidationException(
+              "Privacy group must contain the enclave public key");
+        }
```

Comment:
It feels like this logic could be polished a bit. Maybe separating the logic for finding a ground onchain vs offchain in separate methods would make it a bit easier to read?